unified-scheduler-pool/src/lib.rs

```diff
@@ -194,34 +199,422 @@ impl TaskHandler for DefaultTaskHandler {
+// to be switched (ie. SchedulingContext), just using a single communication channel.
+//
+// Usually, there's no way to prevent one of those threads from mixing current and next contexts
+// while processing messages with a multiple-consumer channel. A condvar or other
+// out-of-bound mechanism is needed to notify about switching of contextual data. That's because
+// there's no way to block those threads reliably on such an switching event just with a channel.
+//
+// However, if the number of consumer can be determined, this can be accomplished just over a
+// single channel, which even carries an in-bound control meta-message with the contexts. The trick
+// is that identical meta-messages as many as the number of threads are sent over the channel,
+// along with new channel receivers to be used (hence the name of _chained_). Then, the receiving
+// thread drops the old channel and is now blocked on receiving from the new channel. In this way,
+// this switching can happen exactly once for each thread.
+//
+// Overall, this greatly simplifies the code, reduces CAS/syscall overhead per messaging to the
+// minimum at the cost of a single heap allocation per switching for the sake of Box-ing the Self
+// type to avoid infinite mem::size_of() due to the recursive type structure. Needless to say, such
+// an allocation can be amortized to be negligible.
+enum ChainedChannel<P1, P2> {
+    Payload(P1),
+    PayloadAndChannel(Box<dyn WithChannelAndPayload<P1, P2>>),
+}
+
+trait WithChannelAndPayload<P1, P2>: Send + Sync {
+    fn payload_and_channel(self: Box<Self>) -> PayloadAndChannelInner<P1, P2>;
+}
+
+type PayloadAndChannelInner<P1, P2> = (P2, Receiver<ChainedChannel<P1, P2>>);
+
+struct PayloadAndChannelWrapper<P1, P2>(PayloadAndChannelInner<P1, P2>);
+
+impl<P1, P2> WithChannelAndPayload<P1, P2> for PayloadAndChannelWrapper<P1, P2>
+where
+    P1: Send + Sync,
+    P2: Send + Sync,
+{
+    fn payload_and_channel(self: Box<Self>) -> PayloadAndChannelInner<P1, P2> {
+        self.0
+    }
+}
+
+impl<P1, P2> ChainedChannel<P1, P2>
+where
+    P1: Send + Sync + 'static,
+    P2: Send + Sync + 'static,
+{
+    fn chain_to_new_channel(payload: P2, receiver: Receiver<Self>) -> Self {
+        Self::PayloadAndChannel(Box::new(PayloadAndChannelWrapper((payload, receiver))))
+    }
+}
+
+fn initialized_result_with_timings() -> ResultWithTimings {
+    (Ok(()), ExecuteTimings::default())
+}
+
 // Currently, simplest possible implementation (i.e. single-threaded)
 // this will be replaced with more proper implementation...
 // not usable at all, especially for mainnet-beta
 #[derive(Debug)]
 pub struct PooledScheduler<TH: TaskHandler> {
     inner: PooledSchedulerInner<Self, TH>,
     context: SchedulingContext,
-    result_with_timings: Mutex<ResultWithTimings>,
 }
 
 #[derive(Debug)]
 pub struct PooledSchedulerInner<S: SpawnableScheduler<TH>, TH: TaskHandler> {
-    id: SchedulerId,
+    thread_manager: ThreadManager<S, TH>,
```

Comment:
note that `ThreadManager` is planned to be wrapped in `Arc<Mutex<_>>` soon. so, i didn't flatten it into PooledSchedulerInner.

Reply:
well well well, I ended up not doing so with different impl.